modules/programs/nnn.nix

```diff
@@ -0,0 +1,132 @@
+
+  pluginModule = types.submodule ({ config, ... }: {
+    options = {
+      src = mkOption {
+        type = with types; nullOr path;
+        example = literalExample ''
+          (pkgs.fetchFromGitHub {
+            owner = "jarun";
+            repo = "nnn";
+            rev = "v4.0";
+            sha256 = "sha256-Hpc8YaJeAzJoEi7aJ6DntH2VLkoR6ToP6tPYn3llR7k=";
+          }) + "/plugins";
+        '';
+        default = null;
+        description = ''
+          Path to the plugin folder.
+        '';
+      };
+
+      mappings = mkOption {
+        type = with types; attrsOf str;
+        description = ''
+          Key mappings to the plugins.
+        '';
+        default = { };
+        example = literalExample ''
+          {
+            c = "fzcd";
+            f = "finder";
+            v = "imgview";
+          };
+        '';
+      };
+    };
+  });
+in {
+  meta.maintainers = with maintainers; [ thiagokokada ];
+
+  options = {
+    programs.nnn = {
+      enable = mkEnableOption "nnn";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.nnn;
+        defaultText = literalExample "pkgs.nnn";
+        example =
+          literalExample "pkgs.nnn.override ({ withNerdIcons = true; });";
+        description = ''
+          Package containing the <command>nnn</command> program.
+        '';
+      };
+
+      finalPackage = mkOption {
+        type = types.package;
+        readOnly = true;
+        visible = false;
+        description = ''
+          Resulting nnn package.
+        '';
+      };
+
+      bookmarks = mkOption {
+        type = with types; attrsOf str;
+        description = ''
+          Directory bookmarks.
+        '';
+        example = literalExample ''
+          {
+            d = "~/Documents";
+            D = "~/Downloads";
+            p = "~/Pictures";
+            v = "~/Videos";
+          };
+        '';
+        default = { };
+      };
+
+      extraPackages = mkOption {
+        type = with types; listOf package;
+        example =
+          literalExample "with pkgs; [ ffmpegthumbnailer mediainfo sxiv ]";
+        description = ''
+          Extra packages available to nnn.
+        '';
+        default = [ ];
+      };
+
+      plugins = mkOption {
+        type = pluginModule;
+        description = ''
+          Manage nnn plugins.
+        '';
+        default = { };
+      };
+    };
+  };
+
+  config = let
+    nnnPackage = cfg.package.overrideAttrs (oldAttrs: {
+      nativeBuildInputs = (oldAttrs.nativeBuildInputs or [ ])
+        ++ [ pkgs.makeWrapper ];
+      postInstall = concatStringsSep "\n" [
+        (oldAttrs.postInstall or "")
+        ''
+          wrapProgram $out/bin/nnn \
+            --prefix PATH : "${makeBinPath cfg.extraPackages}" \
+            --prefix NNN_BMS : "${renderSettings cfg.bookmarks}" \
+            --prefix NNN_PLUG : "${renderSettings cfg.plugins.mappings}"
```

Comment:
This does cause the package to be double wrapped: https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/misc/nnn/default.nix#L49

I tested it and it doesn't seem to be an issue, though.

Reply:
The double wrap:

```
$ cat /etc/profiles/per-user/thiagoko/bin/nnn
#! /nix/store/wv35g5lff84rray15zlzarcqi9fxzz84-bash-4.4-p23/bin/bash -e
export PATH='/nix/store/j2pd4bi3aml8847fvr2dzacrwwcrasw2-bat-0.18.2/bin:/nix/store/zg5ll97l9k09zpdi18hpji97hb0hd5fp-exa-0.10.1/bin:/nix/store/3sb61myhxic6dfbw7hm034rd3jvyjigb-ffmpegthumbnailer-2.2.2/bin:/nix/store/043sbx78dsaas0yqpv7zl544c15sw8xr-mediainfo-21.03/bin:/nix/store/fl4y0by473rgci3h7cmsclmaxc9r5iqa-sxiv-26/bin'${PATH:+':'}$PATH
export NNN_BMS='/:/;D:~/Downloads;d:~/Documents;m:/mnt;p:~/Pictures;v:~/Videos'${NNN_BMS:+':'}$NNN_BMS
export NNN_PLUG='c:fzcd;f:finder;o:fzopen;p:preview-tui;t:nmount;v:imgview'${NNN_PLUG:+':'}$NNN_PLUG
exec -a "$0" "/nix/store/b3rrml5a2x10h6pvdc4a9spn5dy6j28y-nnn-4.0/bin/.nnn-wrapped_"  "$@"

$ cat /nix/store/b3rrml5a2x10h6pvdc4a9spn5dy6j28y-nnn-4.0/bin/.nnn-wrapped_
#! /nix/store/wv35g5lff84rray15zlzarcqi9fxzz84-bash-4.4-p23/bin/bash -e
export PATH='/nix/store/2jh1myphkbxwwnw9zry684lybg3skgs4-file-5.39/bin:/nix/store/781niqj5ddkzxpg6jhh0g6yd3yf4g66p-which-2.21/bin'${PATH:+':'}$PATH
exec -a "$0" "/nix/store/b3rrml5a2x10h6pvdc4a9spn5dy6j28y-nnn-4.0/bin/.nnn-wrapped"  "$@"
```

Reply:
Double wrap is fine in this case imo.

When we finally get the ability to manage wrapping declaratively then the issue will disappear.

Reply:
> When we finally get the ability to manage wrapping declaratively then the issue will disappear.

Is there more information about this (PR, a issue, etc?) so I can read more about it? Quite interested about how this will work.